### PR TITLE
Refactor outgoing packet handling, improves connection speed by 2x

### LINKED ISSUE #2838: Slower connection speed on Outline 1.21.0 Android
### Application

Outline Client

### Describe the bug

We noticed a bug that's slowing down the connection speed of the Outline on certain Android devices with Mediatek processor particularly on Xiaomi Devices. 

We tested this with multiple Xiaomi devices and able to reproduce this slow down.

Our internet speed is great (400Mbps+), there is no kind of censorship / blockage in our test network.

The same server key on iOS gets around 350+Mbps while Android maxing out around 60Mbps down & up. 

This issue is not related WiFi speed of the mobile phone, testing the WiFi itself without VPN gets the full speed of around 350Mbps.

We also tested with different VPN app powered by XRay, we get the maximum speed of around 300Mbps.

Outline - Only 60 Mbps.

There is no issue on iOS/macOS and Windows.

### Steps to reproduce

1. Connect Normally on Xiaomi Devices
2. We confirmed this issue on Poco X7 Pro, Xiaomi Pad Mini, and Xiaomi 13 all running on HyperOS 3.
3. Run Speed Test.
4. Compare with another shadowsocks client like Hiddify, Happ, or any other Shadowsocks supported client.

### What did you expect to happen?

Same download & upload speed as other applications.

### What actually happened?

Speed significantly dropped.

### Outline Version

1.21.0

### What operation system are you using?

Android

### Operating System Version

Multiple Versions

### Screenshots and Videos

_No response_

### PR DESCRIPTION
**Pull request by khantzawhein**
Updated the outgoing packet handling to use a buffered channel for rdBuf, allowing for better performance and backpressure management. Added comments for clarity on the functionality of forwardOutgoingIPPacket and recycling of packet buffers.

This fixes https://github.com/OutlineFoundation/outline-apps/issues/2838, improving peak connection speed on Samsung Galaxy S25 Ultra from 200Mbps to 400Mbps during my testing with speedtest.net.

### CODE REVIEW

**Comment by khantzawhein** [comment]
Before: 

Image

After:

Image

File: network/lwip2transport/device.go
Comment (fortuna) [inline network/lwip2transport/device.go:213]: I need more time to understand this PR, but my impression is that it's introducing a new buffering layer that can actually undermine backpressure.

Backpressure can be currently determined by the sink (the io.Writer here). The buffering is probably in the wrong place. Should be in the sink instead. Otherwise we are adding complexity and violating the single-responsibility design principle.

We need to look into why the sink is blocking. Perhaps lwip is not configured properly? It already has a buffer, it should be using it, but it may be misconfigured. 

Perhaps we cam make a buffered sink wrapper that adds the functionality without adding complexity to this code.

File: network/lwip2transport/device.go [hunk 1]
```diff
@@ -38,11 +38,21 @@ type lwIPDevice struct {
...
 }
 
+// outQueueDepth bounds the outgoing packet queue. At ~2 KiB per slot this caps
+// queue memory at ~512 KiB. Once full, forwardOutgoingIPPacket blocks, which is
+// genuine backpressure (the TUN writer is truly slower) rather than the
+// unconditional per-packet stall it replaces.
+const outQueueDepth = 256
+
+// pktPool recycles the packet copies handed to the TUN writer.
+var pktPool = sync.Pool{New: func() any { b := make([]byte, 0, 2048); return &b }}
```
Comment (fortuna) [inline network/lwip2transport/device.go]: We should avoid globals. It's always a pain. Better to add to lwipDevice instead. But see my comments below. We may need a different approach.

**Review by fortuna — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution. That's a significant performance increase. Do you know if that's a new issue?

File: network/lwip2transport/device.go
Comment (fortuna) [reply network/lwip2transport/device.go:213]: @greptile investigate the sink and why it blocks

**Comment by khantzawhein** [comment]
> Thanks a lot for the contribution. That's a significant performance increase. Do you know if that's a new issue?

No, I tested with old 1.19.0 version and it still has performance issue.

Most of our users are not aware of this because they don't usually have internet speed faster than 100Mbps, and it only has visible impact on Android. But some Xiaomi devices on Mediatek chipsets have huge performance impact with speed typically capped to <60Mbps when connected. (I have verified that, with this PR, there is no performance issue on these devices anymore.)

Since I am quite busy this week, I will review the PR a few days later, or you can create a new PR and close mine, so your team can ship it faster into new version of Outline.

**Comment by khantzawhein** [comment]
@greptile check the PR again.

**Comment by khantzawhein** [comment]
@greptile check the PR again.

**Comment by khantzawhein** [comment]
@greptile check the PR again.

File: network/lwip2transport/device.go
Comment (fortuna) [reply network/lwip2transport/device.go:243]: @khantzawhein please don't resolve the conversations without addressing it.

The change is in the wrong layer. We should wrap the sink instead. Also, use net.FileConn for the fd.

In fact, I would suggest making only the net.FileConn change. That alone may solve the problem. If not, then wrap it with the buffered write closer.

**Review by fortuna — CHANGES_REQUESTED** [review]
The layering issue still needs to be addressed.
